app/application/use_cases/authenticate_user.py

```
@@ -1 +1,37 @@
+
+class AuthenticateUser:
+    def __init__(self, repo: UserRepository, hasher: PasswordHasher):
+        self._repo = repo
+        self._hasher = hasher
+
+    def execute(self, data: LoginInput) -> User:
+        user = self._repo.get_by_email(data.email)
+        if not user:
+            # Timing attack mitigation (verify a fake hash)? 
+            # For now, simple return to avoid complexity over-engineering
+            raise InvalidCredentialsError("Invalid email or password")
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)